Rename the GD7965 driver to UC81xx and add support for UC8176

andysan opened the pull request:
The GD7965 is really a driver for an UltraChip UC8179 controller. There are also EPD profiles in-tree that use the UC8176 controller which is mostly compatible (some registers have a slightly different layout).

This pull requests renames the driver to UC81xx to reflect the fact that it supports multiple UltraChip devices and it includes a number of changes to implement proper support for the UC8176. These changes should normally be included in two different pull requests, but I would like to avoid the risk of those PRs spanning two releasees since they would both affect the DT bindings.

The main user-facing change is the change of compatible strings. The gooddisplay,gd7965 has been replaced by chip-specific compatible strings (ultrachip,uc8176 and ultrachip,uv8179). There is no longer a generic driver string since the different chips need different quirks.

andysan committed:
drivers: Rename GD7965 to UC81xx

The GD7965 driver is really just a vendor name for the UltraChip
UC8179. Rename the driver to UC81xx since there are other chips in the
family (e.g., the UC8176) with an almost identical register interface.

Signed-off-by: Andreas Sandberg <andreas@sandberg.uk>

andysan committed:
drivers: uc81xx: Use device-specific compatibles

The UC8176 and UC8179 chips that exist in tree have subtly different
register layouts. Use separate compatible strings for these chips and
a quirks structure that describe device-specific behavior.

Signed-off-by: Andreas Sandberg <andreas@sandberg.uk>

andysan committed:
drivers: uc81xx: Fix incorrect UC8176 CDI layout

The CDI register in UC8176 and UC8179 have different layouts. Add a
helper function to the quirks structure to handle CDI updates.

Signed-off-by: Andreas Sandberg <andreas@sandberg.uk>

andysan committed:
drivers: uc81xx: Avoid using auto sequencing

The UC8179 supports automatic power management where the DC-DC is
enabled automatically. This is not supported on the UC8176. Explicitly
call PON/DRF/POF from uc81xx_update_display instead. This ensures that
the DC-DC is only enabled while actually updating the display for all
supported chips.

Signed-off-by: Andreas Sandberg <andreas@sandberg.uk>

mbolivar-nordic commented on tests/drivers/build_all/sensor/spi.dtsi:
```diff
@@ -80,8 +80,20 @@ test_spi_lpd8806: lpd8806@9 {
 
+test_spi_uc8176: uc8176@a {
+	compatible = "ultrachip,uc8176";
+	label = "UC8176";
```
Remove this; we are phasing out use of label properties in upstream Zephyr, so let's not add any more, please.

andysan replied:
Done.

mbolivar-nordic commented on drivers/display/uc81xx.c:
```diff
@@ -279,13 +274,9 @@ static int uc81xx_write(const struct device *dev, const uint16_t x, const uint16
-					   data->bdd_polarity)) {
-			return -EIO;
-		}
+	/* Enable boarder output */
```
Did you mean "border?" Here and above.

andysan replied:
Done.

mbolivar-nordic commented on drivers/display/uc81xx.c:
```diff
@@ -279,13 +274,9 @@ static int uc81xx_write(const struct device *dev, const uint16_t x, const uint16
-		}
+	/* Enable boarder output */
+	if (config->quirks->set_cdi(dev, true)) {
+		return -EIO;
```
Why swallow the error instead of propagating it up? Here and above.

mbolivar-nordic commented on drivers/display/uc81xx.c:
```diff
@@ -431,17 +421,8 @@ static int uc81xx_controller_init(const struct device *dev)
-		}
-	}
+	/* Set CDI and enable border output */
+	config->quirks->set_cdi(dev, true);
```
No error handling?

mbolivar-nordic reviewed (changes requested):
Overall looks quite nice. I think @jfischer-no should be the assignee however.

A few minor requests and questions from me.

andysan commented on boards/shields/waveshare_epaper/waveshare_epaper_gdew042t2.overlay:
```diff
@@ -7,14 +7,14 @@
-    		label = "GD7965";
+	uc8176: uc8176@0 {
+		compatible =  "gooddisplay,gdew042t2", "ultrachip,uc81xx";
+		label = "UC8176";
```
I just had a slightly annoying realisation. The LVGL integration still relies on labels in a few places so I think we'll need to keep these for now. :(

andysan commented:
@mbolivar-nordic: I have rebased this change and fixed a couple of merge conflicts. Other than the merge conflicts, the change should be identical to when @jfischer-no approved it last week. Are we OK to merge this or do you want someone else to have a look?

mbolivar-nordic reviewed (approved):
Blindly reapproving to get you one of the +1s you need.

We need @jfischer-no to reapprove again as well due to our github repository configuration.

andysan commented:
Are we ok to merge this? I have some changes (fast partial refresh) that depend on this.
